src/jupyterlabcontroller/models/v1/external/prepuller_config.py

```diff
@@ -0,0 +1,173 @@
+            " for lab image."
+        ),
+    )
+    name: str = Field(
+        ...,
+        title="name",
+        example="Latest Daily (Daily 2077_10_23)",
+        description=("Human-readable version of image tag"),
+    )
+
+
+class PrepullerConfig(BaseModel):
+    """See https://sqr-059.lsst.io for how this is used."""
+
+    registry: str = Field(
+        "registry.hub.docker.com",
+        title="registry",
+        example="lighthouse.ceres",
+        description="Hostname of Docker repository",
+    )
+    docker: Optional[DockerDefinition] = None
+    gar: Optional[GARDefinition] = None
+    recommendedTag: str = Field(
+        "recommended",
+        title="recommended",
+        example="recommended",
+        description="Image tag to use as `recommended` image",
+    )
+    numReleases: int = Field(
+        1,
+        title="numReleases",
+        example=1,
+        description="Number of Release images to prepull and display in menu.",
+    )
+    numWeeklies: int = Field(
+        2,
+        title="numWeeklies",
+        example=2,
+        description="Number of Weekly images to prepull and display in menu.",
+    )
+    numDailies: int = Field(
+        3,
+        title="numDailies",
+        example=3,
+        description="Number of Daily images to prepull and display in menu.",
+    )
+    cycle: Optional[int] = Field(
+        None,
+        title="cycle",
+        example=27,
+        description=(
+            "Cycle number describing XML schema version of this"
+            " image.  Currently only used by T&S RSP."
+        ),
+    )
+    pin: Optional[List[ImagePathAndName]] = Field(
+        None,
+        title="pin",
+        example=["lighthouse.ceres/library/sketchbook:d_2077_10_23"],
+        description=(
+            "List of images to prepull and pin to the menu, "
+            "even if they would not normally be prepulled."
+        ),
+    )
+    aliasTags: List[str] = Field(
+        [],
+        title="aliasTags",
+        example=["recommended_cycle0027"],
+        description="Additional alias tags for this instance.",
+    )
+
+    @root_validator
+    def registry_defined(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        for x in values.keys():
+            if x == "gar" or x == "docker":
+                return values
+        assert False, "Exactly one of 'docker' or 'gar' must be defined"
+
+    @validator("registry")
+    def validate_registry(cls, v: str) -> str:
+        # only here to ensure that registry is validated for the GAR
+        # validator
+        return v
+
+    @validator("gar")
+    def gar_registry_host(
+        cls, v: GARDefinition, values: Dict[str, str]
+    ) -> GARDefinition:
+        reg = values["registry"]
+        gsuf = "-docker.pkg.dev"
+        if v.location != "":
+            assert v.location == f"{reg}-{gsuf}"
+        else:
+            v.location = f"{reg}-{gsuf}"
+        return v
```

Comment:
How would I do that?  If v.location isn't set, I want to set it appropriately, but if v.location is set I want to ensure that it matches the location that comes from appending the rest of the domain to values["registry"] ; I see how I would match the correct suffix with a regex, but not both.

Reply:
Oh, I see, it depends on another parameter and I wasn't reading that right.

This is pretty awkward too, since you have to validate a sub-model and alter it on the fly. I think that part is is my fault for lifting the `registry` attribute up into the `images` definition because it was common to both `gar` and `docker`. I think the right fix is to just duplicate `registry` in both the `gar` and `docker` models, and then all of this awkwardness goes away because you don't need to set a default.

But yeah, I don't think you can avoid a validation function in the `GARDefinition` model, since the default depends on another field.

Reply:
I have it working OK, and "registry" and "repository" work as access properties top-level.